File: y11459/src/entities/AsyncTask.ts
import { Entity, Column, ManyToOne, JoinColumn } from 'typeorm';
import { BaseEntity } from './BaseEntity';
import { Batch } from './Batch';

export type TaskStatus = 'pending' | 'processing' | 'retry_waiting' | 'manual_waiting' | 'failed' | 'completed';
export type TaskType = 'import' | 'match' | 'export' | 'calculate' | 'notify';

@Entity('async_tasks')
export class AsyncTask extends BaseEntity {
  @Column({ name: 'batch_id', nullable: true })
  batchId: string;

  @ManyToOne(() => Batch, batch => batch.asyncTasks, { nullable: true })
  @JoinColumn({ name: 'batch_id' })
  batch: Batch;

  @Column({
    type: 'simple-enum',
    enum: ['import', 'match', 'export', 'calculate', 'notify'],
    default: 'import'
  })
  taskType: TaskType;

  @Column({
    type: 'simple-enum',
    enum: ['pending', 'processing', 'retry_waiting', 'manual_waiting', 'failed', 'completed'],
    default: 'pending'
  })
  status: TaskStatus;

  @Column({ name: 'retry_count', default: 0 })
  retryCount: number;

  @Column({ name: 'max_retries', default: 3 })
  maxRetries: number;

  @Column({ name: 'next_retry_at', nullable: true, type: 'datetime' })
  nextRetryAt: Date;

  @Column({ name: 'started_at', nullable: true, type: 'datetime' })
  startedAt: Date;

  @Column({ name: 'completed_at', nullable: true, type: 'datetime' })
  completedAt: Date;

  @Column({ name: 'failed_at', nullable: true, type: 'datetime' })
  failedAt: Date;

  @Column({ name: 'progress', default: 0 })
  progress: number;

  @Column({ name: 'total_items', default: 0 })
  totalItems: number;

  @Column({ name: 'processed_items', default: 0 })
  processedItems: number;

  @Column({ name: 'error_message', nullable: true, type: 'text' })
  errorMessage: string;

  @Column({ name: 'error_stack', nullable: true, type: 'text' })
  errorStack: string;

  @Column({ name: 'error_code', nullable: true })
  errorCode: string;

  @Column({ name: 'assigned_to', nullable: true })
  assignedTo: string;

  @Column({ name: 'resolution_note', nullable: true, type: 'text' })
  resolutionNote: string;

  @Column({ type: 'simple-json', nullable: true })
  payload: Record<string, any>;

  @Column({ type: 'simple-json', nullable: true })
  result: Record<string, any>;
}

File: y11505/app/models/async_task.py
from sqlalchemy import Column, String, DateTime, Text, Boolean, Integer, JSON
from datetime import datetime

from app.core.database import Base
from app.core.constants import TaskStatus


class AsyncTask(Base):
    __tablename__ = "async_tasks"

    id = Column(String, primary_key=True, index=True)
    task_name = Column(String, nullable=False, index=True)
    task_type = Column(String, nullable=False)
    
    batch_id = Column(String, nullable=True, index=True)
    record_id = Column(String, nullable=True)
    
    status = Column(String, default=TaskStatus.PENDING, index=True)
    
    payload = Column(JSON, nullable=True)
    result = Column(JSON, nullable=True)
    
    error_message = Column(Text, nullable=True)
    error_trace = Column(Text, nullable=True)
    
    retry_count = Column(Integer, default=0)
    max_retries = Column(Integer, default=3)
    
    last_run_time = Column(DateTime, nullable=True)
    next_run_time = Column(DateTime, nullable=True)
    
    operator = Column(String, nullable=True)
    
    is_cancelled = Column(Boolean, default=False)
    
    created_at = Column(DateTime, default=datetime.now)
    updated_at = Column(DateTime, default=datetime.now, onupdate=datetime.now)
